### PR TITLE
fix: respect default and last direction on new line and indent

### PR DESCRIPTION
**Pull request by zoli**

### CODE REVIEW

File: lib/src/editor/block_component/base_component/text_direction_mixin.dart [hunk 1]
```diff
@@ -106,6 +95,51 @@ TextDirection calculateNodeDirection({
...
+    prevOrParentNodeDirection =
+        _getDirectionFromNode(node.previous!, defaultTextDirection);
+  }
+  if (node.parent != null && prevOrParentNodeDirection == null) {
+    prevOrParentNodeDirection =
+        _getDirectionFromNode(node.parent!, defaultTextDirection);
+  }
+
+  return prevOrParentNodeDirection;
+}
```
Comment (LucasXu0) [inline lib/src/editor/block_component/base_component/text_direction_mixin.dart]:
```suggestion
TextDirection? _getDirectionFromPreviousOrParentNode(
  Node node,
  String? defaultTextDirection,
) {
  if (node.previous != null) {
    return _getDirectionFromNode(node.previous!, defaultTextDirection);
  }
  if (node.parent != null) {
    return _getDirectionFromNode(node.parent!, defaultTextDirection);
  }
  return null;
}
```

File: lib/src/editor/block_component/base_component/text_direction_mixin.dart [hunk 2]
```diff
@@ -106,6 +95,51 @@ TextDirection calculateNodeDirection({
 
   // if the value is auto and the text isn't null or empty,
   // calculate the text direction by the text
+  final direction = _determineTextDirection(text);
+  if (direction != null) {
+    return direction;
+  }
+
+  return defaultTextDirection?.toTextDirection() ?? layoutDirection;
```
Comment (LucasXu0) [inline lib/src/editor/block_component/base_component/text_direction_mixin.dart]:
```suggestion
  return _determineTextDirection(text) ??
      defaultTextDirection?.toTextDirection() ??
      layoutDirection;
```

File: lib/src/editor/block_component/base_component/text_direction_mixin.dart [hunk 3]
```diff
@@ -106,6 +95,51 @@ TextDirection calculateNodeDirection({
...
+  if (nodeDirection != null) {
+    if (nodeDirection == blockComponentTextDirectionAuto) {
+      return node.selectable?.textDirection();
+    } else {
+      return nodeDirection.toTextDirection();
+    }
+  }
+
+  return null;
+}
```
Comment (LucasXu0) [inline lib/src/editor/block_component/base_component/text_direction_mixin.dart]:
```suggestion
TextDirection? _getDirectionFromNode(Node node, String? defaultTextDirection) {
  final nodeDirection = node.direction(
    defaultTextDirection == blockComponentTextDirectionAuto
        ? blockComponentTextDirectionAuto
        : null,
  );
  if (nodeDirection == blockComponentTextDirectionAuto) {
    return node.selectable?.textDirection();
  } else {
    return nodeDirection?.toTextDirection();
  }
}
```

File: lib/src/editor/block_component/base_component/text_direction_mixin.dart [hunk 4]
```diff
@@ -114,23 +148,14 @@ TextDirection calculateNodeDirection({
...
-      node.attributes[blockComponentTextDirection] != null;
-
-  if (textDirectionCheck(node.previous)) {
-    return node.previous;
-  } else if (textDirectionCheck(node.parent)) {
-    return node.parent;
+extension on Node {
+  String? direction(String? defaultDirection) {
+    return (attributes[blockComponentTextDirection] as String?) ??
+        defaultDirection;
```
Comment (LucasXu0) [inline lib/src/editor/block_component/base_component/text_direction_mixin.dart]:
```suggestion
extension on Node {
  String? direction(String? defaultDirection) =>
      attributes[blockComponentTextDirection] as String? ?? defaultDirection;
```

File: lib/src/editor/editor_component/service/renderer/block_component_widget.dart [hunk 5]
```diff
@@ -77,19 +77,18 @@ mixin NestedBlockComponentStatefulWidgetMixin<
...
     if (node.children.isNotEmpty) {
-      TextDirection? lastDirection;
-      if (node.children.first.key.currentState != null) {
-        lastDirection = (node.children.first.key.currentState
-                as BlockComponentTextDirectionMixin)
-            .lastDirection;
+      final firstChild = node.children.first;
+      final currentState =
+          firstChild.key.currentState as BlockComponentTextDirectionMixin?;
+      if (currentState != null) {
```
Comment (zoli) [inline lib/src/editor/editor_component/service/renderer/block_component_widget.dart:83]: Isn't there a case when the `currentState` is null and we still want to do the calculation?
last direction is optional parameter.

**Pull request merged**